src/gcp.rs

```diff
@@ -194,21 +199,28 @@ impl GcpAuthenticator {
     async fn token_regenerator(self, sender: watch::Sender<()>) {
         match self {
             Self::Credentials(inner) => {
-                let period =
-                    Duration::from_secs(inner.token.read().unwrap().expires_in() as u64 / 2);
-                let mut interval = tokio::time::interval_at(Instant::now() + period, period);
+                let expires_in = inner.token.read().unwrap().expires_in() as u64;
+                let mut deadline = Duration::from_secs(expires_in - METADATA_TOKEN_EXPIRY_MARGIN_SECS);
+                let mut next_refresh = tokio::time::sleep(deadline);
+                next_refresh.await;
                 loop {
```

Comment:
Very minor nitpick, but this can be simplified somewhat:

```suggestion
                loop {
                    tokio::time::sleep(deadline).await;
```